zebra-chain/src/sapling/tree.rs

```diff
@@ -72,37 +82,181 @@ impl fmt::Debug for Root {
     }
 }
 
-impl From<SaplingNoteCommitmentTree> for Root {
-    fn from(_tree: SaplingNoteCommitmentTree) -> Self {
-        // TODO: The Sapling note commitment tree requires a Pedersen
-        // hash function, not SHA256.
-
-        // let mut hash_writer = Sha256dWriter::default();
-        // sapling_note_commitment_tree
-        //     .zcash_serialize(&mut hash_writer)
-        //     .expect("A Sapling note commitment tree must serialize.");
-        // Self(hash_writer.finish())
+/// Sapling Note Commitment Tree
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+struct NoteCommitmentTree {
+    /// The root node of the tree (often used as an anchor).
+    root: Root,
+    /// The height of the tree (maximum height for Sapling is 32).
+    height: u8,
+    /// The number of leaves (note commitments) in this tree.
+    count: u32,
+}
 
+impl From<Vec<NoteCommitment>> for NoteCommitmentTree {
+    fn from(_values: Vec<NoteCommitment>) -> Self {
         unimplemented!();
     }
 }
 
-impl SaplingNoteCommitmentTree {
-    /// Get the Jubjub-based Pedersen hash of root node of this merkle
-    /// tree of commitment notes.
-    pub fn hash(&self) -> [u8; 32] {
-        unimplemented!();
+impl From<Vec<jubjub::Fq>> for NoteCommitmentTree {
+    fn from(values: Vec<jubjub::Fq>) -> Self {
+        if values.is_empty() {
+            return NoteCommitmentTree {
+                root: Root::default(),
+                height: 0,
+                count: 0,
+            };
+        }
+
+        let count = values.len() as u32;
+        let mut height = 0u8;
+        let mut current_layer: Vec<[u8; 32]> =
+            values.into_iter().map(|cm_u| cm_u.to_bytes()).collect();
+
+        while usize::from(height) < MERKLE_DEPTH {
```

Comment:
it looks like `height` us used as a `usize` in a few places but as a u8 in only one. Is it worth changing this to usize by default?

Reply:
I think I might want a `Height` type that will have constraints that it is always <= SAPLING_MERKLE_DEPTH (32), which is kinda why I'm keeping it u8 everywhere and not just `usize` because usize varies by platform